pkg/util/ranger/types.go

```diff
@@ -329,34 +394,124 @@ func (rs Ranges) Subset(tc types.Context, superRanges Ranges) bool {
+	if !checkCollators(ran, otherRange, len(otherRange.LowVal)) {
+		return false
+	}
+
+	// Either otherRange is closed or both ranges have the same open/close setting.
+	lowExcludeOK := !otherRange.LowExclude || ran.LowExclude == otherRange.LowExclude
+	highExcludeOK := !otherRange.HighExclude || ran.HighExclude == otherRange.HighExclude
 	if !lowExcludeOK || !highExcludeOK {
 		return false
 	}
 
-	return prefix(tc, superRange.LowVal, ran.LowVal, len(superRange.LowVal), ran.Collators) &&
-		prefix(tc, superRange.HighVal, ran.HighVal, len(superRange.LowVal), ran.Collators)
+	return prefix(tc, otherRange.LowVal, ran.LowVal, len(otherRange.LowVal), ran.Collators) &&
+		prefix(tc, otherRange.HighVal, ran.HighVal, len(otherRange.LowVal), ran.Collators)
+}
+
+// IntersectRange computes intersection between two ranges. err is set of something went wrong
+// during comparison.
+func (ran *Range) IntersectRange(tc types.Context, otherRange *Range) (*Range, error) {
+	intersectLength := max(len(ran.LowVal), len(otherRange.LowVal))
+	result := &Range{
+		LowVal:    make([]types.Datum, 0, intersectLength),
+		HighVal:   make([]types.Datum, 0, intersectLength),
+		Collators: make([]collate.Collator, 0, intersectLength),
+	}
+
+	if len(ran.LowVal) > len(otherRange.LowVal) {
+		result.Collators = ran.Collators
+	} else {
+		result.Collators = otherRange.Collators
+	}
+
+	lowVsHigh, err := compareLexicographically(tc, ran.LowVal, otherRange.HighVal, ran.Collators,
+		ran.LowExclude, otherRange.HighExclude, true, false)
+	if err != nil {
+		return &Range{}, err
+	}
+	if lowVsHigh == 1 {
+		return nil, nil
+	}
+
+	lowVsHigh, err = compareLexicographically(tc, otherRange.LowVal, ran.HighVal, ran.Collators,
+		otherRange.LowExclude, ran.HighExclude, true, false)
+	if err != nil {
+		return &Range{}, err
+	}
+	if lowVsHigh == 1 {
+		return nil, nil
+	}
+
+	lowVsLow, err := compareLexicographically(tc, ran.LowVal, otherRange.LowVal,
+		ran.Collators, ran.LowExclude, otherRange.LowExclude, true, true)
+	if err != nil {
+		return &Range{}, err
+	}
+	if lowVsLow == -1 {
+		result.LowVal = otherRange.LowVal
+		result.LowExclude = otherRange.LowExclude
+	} else {
+		result.LowVal = ran.LowVal
+		result.LowExclude = ran.LowExclude
+	}
+
+	highVshigh, err := compareLexicographically(tc, ran.HighVal, otherRange.HighVal,
+		ran.Collators, ran.HighExclude, otherRange.HighExclude, false, false)
+	if err != nil {
+		return &Range{}, err
+	}
+	if highVshigh == 1 {
+		result.HighVal = otherRange.HighVal
+		result.HighExclude = otherRange.HighExclude
+	} else {
+		result.HighVal = ran.HighVal
+		result.HighExclude = ran.HighExclude
+	}
+	return result, nil
+}
+
+// IntersectRanges computes pairwise intersection between each element in rs and otherRangeList.
+func (rs Ranges) IntersectRanges(tc types.Context, otherRanges Ranges) Ranges {
+	result := Ranges{}
+	for _, rsRange := range rs {
+		for _, otherRange := range otherRanges {
```

Comment:
Ideally, we can order the two `Ranges` and avoid calculating the intersection for every pair between them, right?

Reply:
Correct and that will be done in subsequent PR when we optimize the intersections and remove the flag.